src/crimson/net/Socket.h

```diff
@@ -136,7 +125,176 @@ class Socket
+  std::optional<seastar::server_socket> listener;
+  seastar::gate shutdown_gate;
+
+  using sharded_service_t = seastar::sharded<FixedCPUServerSocket>;
+  std::unique_ptr<sharded_service_t> service;
+
+  struct construct_tag {};
+
+  static seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_ms);
+  }
+
+  seastar::future<> reset() {
+    return container().invoke_on_all([] (auto& ss) {
+      assert(ss.shutdown_gate.is_closed());
+      ss.shutdown_gate = seastar::gate();
+      ss.addr = entity_addr_t();
+      ss.listener.reset();
+    });
+  }
+
+public:
+  FixedCPUServerSocket(seastar::shard_id cpu, construct_tag) : cpu{cpu} {}
+  ~FixedCPUServerSocket() {
+    assert(!listener);
+    // detect whether user have called destroy() properly
+    ceph_assert(!service);
+  }
+
+  FixedCPUServerSocket(FixedCPUServerSocket&&) = delete;
+  FixedCPUServerSocket(const FixedCPUServerSocket&) = delete;
+  FixedCPUServerSocket& operator=(const FixedCPUServerSocket&) = delete;
+
+  seastar::future<bool> listen(entity_addr_t addr) {
+    assert(seastar::engine().cpu_id() == cpu);
+    logger().debug("FixedCPUServerSocket::listen({})...", addr);
+    return container().invoke_on_all([addr] (auto& ss) {
+      ss.addr = addr;
+      seastar::socket_address s_addr(addr.in4_addr());
+      seastar::listen_options lo;
+      lo.reuse_address = true;
+      lo.set_fixed_cpu(ss.cpu);
+      ss.listener = seastar::listen(s_addr, lo);
+    }).then([] {
+      return true;
+    }).handle_exception_type([addr] (const std::system_error& e) {
+      if (e.code() == error::address_in_use) {
+        logger().debug("FixedCPUServerSocket::listen({}): address in use", addr);
+        return false;
+      } else {
+        throw;
+      }
+    }).handle_exception([addr] (auto eptr) {
+      logger().error("FixedCPUServerSocket::listen({}): "
+                     "got unexpeted error {}", addr, eptr);
+      ceph_abort();
+      return false;
+    });
+  }
+
+  // fn_accept should be a nothrow function of type
+  // seastar::future<>(SocketRef, entity_addr_t)
+  template <typename Func>
+  seastar::future<> accept(Func&& fn_accept) {
+    assert(seastar::engine().cpu_id() == cpu);
+    logger().debug("FixedCPUServerSocket({})::accept()...", addr);
+    return container().invoke_on_all(
+        [fn_accept = std::move(fn_accept)] (auto& ss) mutable {
+      assert(ss.listener);
+      // gate accepting
+      // FixedCPUServerSocket::shutdown() will drain the continuations in the gate
+      // so ignore the returned future
+      std::ignore = seastar::with_gate(ss.shutdown_gate,
+          [&ss, fn_accept = std::move(fn_accept)] () mutable {
+        return seastar::keep_doing([&ss, fn_accept = std::move(fn_accept)] () mutable {
+          return ss.listener->accept().then(
+              [&ss, fn_accept = std::move(fn_accept)]
+              (seastar::accept_result accept_result) mutable {
+            if (seastar::engine().cpu_id() == ss.cpu) {
```

Comment:
as discussed offline, better off putting
```c++
// set_fixed _cpu() failed to work!
assert(seastar::engine().cpu_id() == ss.cpu);
```

Reply:
fixed in https://github.com/ceph/ceph/pull/32632/commits/c16013b3751e3e9591f7a532547c1b5a87198f2d